File: src-tauri/src/security.rs
use aes_gcm::{
    aead::{Aead, KeyInit},
    Aes256Gcm, Nonce,
};
use base64::{engine::general_purpose::STANDARD as BASE64, Engine};
use rand::Rng;
use sha2::{Digest, Sha256};
use std::sync::OnceLock;

static ENCRYPTION_KEY: OnceLock<[u8; 32]> = OnceLock::new();

/// 获取稳定的机器标识（不依赖 IP 地址）
fn get_machine_id() -> String {
    let mut id = String::new();

    // macOS: 使用 IOPlatformUUID（硬件唯一标识，不随网络变化）
    #[cfg(target_os = "macos")]
    {
        let mut found_uuid = false;
        if let Ok(output) = std::process::Command::new("ioreg")
            .args(["-rd1", "-c", "IOPlatformExpertDevice"])
            .output()
        {
            let stdout = String::from_utf8_lossy(&output.stdout);
            if let Some(uuid_start) = stdout.find("IOPlatformUUID") {
                let rest = &stdout[uuid_start..];
                if let Some(quote_start) = rest.find('"') {
                    let uuid_part = &rest[quote_start + 1..];
                    if let Some(quote_end) = uuid_part.find('"') {
                        let uuid = &uuid_part[..quote_end];
                        id.push_str(uuid);
                        found_uuid = true;
                    }
                }
            }
        }
        // 如果 IOPlatformUUID 获取失败，回退到主机名（与旧行为一致，确保密钥稳定性）
        if !found_uuid {
            id.push_str(
                &hostname::get()
                    .map(|h| h.to_string_lossy().to_string())
                    .unwrap_or_else(|_| "default-host".to_string()),
            );
        }
    }

    // 其他平台: 使用主机名（通常是稳定的计算机名，而非 IP）
    #[cfg(not(target_os = "macos"))]
    {
        id.push_str(
            &hostname::get()
                .map(|h| h.to_string_lossy().to_string())
                .unwrap_or_else(|_| "default-host".to_string()),
        );
    }

    // 用户名
    id.push_str(&std::env::var("USER").unwrap_or_else(|_| {
        std::env::var("USERNAME").unwrap_or_else(|_| "default-user".to_string())
    }));

    // 应用标识（避免多应用密钥冲突）
    id.push_str("i-dblink");

    id
}

fn get_key() -> &'static [u8; 32] {
    ENCRYPTION_KEY.get_or_init(|| {
        let machine_id = get_machine_id();
        let mut hasher = Sha256::new();
        hasher.update(machine_id.as_bytes());
        let result = hasher.finalize();
        let mut key = [0u8; 32];
        key.copy_from_slice(&result);
        key
    })
}

pub fn encrypt_password(password: &str) -> Result<String, String> {
    let key = get_key();
    let cipher = Aes256Gcm::new(key.into());

    let mut nonce_bytes = [0u8; 12];
    rand::thread_rng().fill(&mut nonce_bytes);
    let nonce = Nonce::from_slice(&nonce_bytes);

    let ciphertext = cipher
        .encrypt(nonce, password.as_bytes())
        .map_err(|e| format!("Encryption failed: {}", e))?;

    let mut result = nonce_bytes.to_vec();
    result.extend(ciphertext);
    Ok(BASE64.encode(&result))
}

pub fn decrypt_password(encrypted: &str) -> Result<String, String> {
    let data = BASE64
        .decode(encrypted)
        .map_err(|e| format!("Invalid base64: {}", e))?;

    if data.len() < 12 {
        return Err("Invalid encrypted data".to_string());
    }

    let (nonce_bytes, ciphertext) = data.split_at(12);
    let nonce = Nonce::from_slice(nonce_bytes);

    let key = get_key();
    let cipher = Aes256Gcm::new(key.into());

    let plaintext = cipher
        .decrypt(nonce, ciphertext)
        .map_err(|e| format!("Decryption failed: {}", e))?;

    String::from_utf8(plaintext).map_err(|e| format!("Invalid UTF-8: {}", e))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_encrypt_decrypt() {
        let password = "my_secret_password";
        let encrypted = encrypt_password(password).unwrap();
        let decrypted = decrypt_password(&encrypted).unwrap();
        assert_eq!(password, decrypted);
    }

    #[test]
    fn test_same_machine_same_key() {
        let key1 = get_key();
        let key2 = get_key();
        assert_eq!(key1, key2);
    }
}

File: backend/security.go
package backend

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"io"
	"os"
	"os/exec"
	"runtime"
	"strings"
	"sync"

	"golang.org/x/crypto/scrypt"
)

var (
	keyOnce sync.Once
	key     []byte
	keyErr  error
)

// getMachineID 获取机器标识
func getMachineID() string {
	var id strings.Builder

	// macOS: 使用 IOPlatformUUID
	if runtime.GOOS == "darwin" {
		if out, err := exec.Command("ioreg", "-rd1", "-c", "IOPlatformExpertDevice").Output(); err == nil {
			stdout := string(out)
			if idx := strings.Index(stdout, "IOPlatformUUID"); idx != -1 {
				rest := stdout[idx:]
				if start := strings.Index(rest, `"`); start != -1 {
					uuidPart := rest[start+1:]
					if end := strings.Index(uuidPart, `"`); end != -1 {
						id.WriteString(uuidPart[:end])
					}
				}
			}
		}
	}

	// 如果 macOS UUID 没获取到，或其他平台，使用 hostname
	if id.Len() == 0 {
		if hostname, err := os.Hostname(); err == nil {
			id.WriteString(hostname)
		} else {
			id.WriteString("default-host")
		}
	}

	// 用户名
	if user := os.Getenv("USER"); user != "" {
		id.WriteString(user)
	} else if user := os.Getenv("USERNAME"); user != "" {
		id.WriteString(user)
	} else {
		id.WriteString("default-user")
	}

	// 应用标识（作为 salt 的一部分）
	id.WriteString("i-dblink")

	return id.String()
}

// getKey 获取加密密钥（32 字节）。
// 用 scrypt KDF（N=2^16, r=8, p=1）从 machineID 派生密钥，
// 替代之前的单次 SHA256，大幅提高离线暴力破解成本。
// salt = SHA256(machineID)，使每个机器的密钥互不相同。
func getKey() ([]byte, error) {
	keyOnce.Do(func() {
		machineID := getMachineID()
		salt := sha256.Sum256([]byte(machineID + "|salt"))
		// scrypt 不允许在启动时失败；若失败（极端内存限制）记录错误，调用方返回错误
		key, keyErr = scrypt.Key([]byte(machineID), salt[:], 1<<16, 8, 1, 32)
	})
	return key, keyErr
}

// getKeyLegacy 返回旧的密钥派生（单次 SHA256），仅用于解密历史已存的密码。
// 新加密一律走 scrypt 的 getKey。
func getKeyLegacy() []byte {
	machineID := getMachineID()
	hash := sha256.Sum256([]byte(machineID))
	return hash[:]
}

func decryptWith(encrypted string, k []byte) (string, error) {
	data, err := base64.StdEncoding.DecodeString(encrypted)
	if err != nil {
		return "", fmt.Errorf("invalid base64: %w", err)
	}
	block, err := aes.NewCipher(k)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	nonceSize := gcm.NonceSize()
	if len(data) < nonceSize {
		return "", fmt.Errorf("invalid encrypted data")
	}
	nonce, ciphertext := data[:nonceSize], data[nonceSize:]
	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return "", err
	}
	return string(plaintext), nil
}

// EncryptPassword 加密密码
func EncryptPassword(password string) (string, error) {
	k, err := getKey()
	if err != nil {
		return "", fmt.Errorf("failed to derive key: %w", err)
	}

	block, err := aes.NewCipher(k)
	if err != nil {
		return "", fmt.Errorf("failed to create cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("failed to create GCM: %w", err)
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return "", fmt.Errorf("failed to generate nonce: %w", err)
	}

	ciphertext := gcm.Seal(nonce, nonce, []byte(password), nil)
	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

// DecryptPassword 解密密码。先用 scrypt 新密钥，失败则回退到旧的 SHA256 密钥
// （向后兼容：用户已有的加密密码仍可解密，解密后由调用方重新加密即可）。
func DecryptPassword(encrypted string) (string, error) {
	k, err := getKey()
	if err == nil {
		if pt, e := decryptWith(encrypted, k); e == nil {
			return pt, nil
		}
	}
	// fallback：旧密钥派生（历史数据）
	pt, err := decryptWith(encrypted, getKeyLegacy())
	if err != nil {
		return "", fmt.Errorf("decryption failed: %w", err)
	}
	return pt, nil
}
